### ISSUE TITLE
How to keep the body scripts within the body?

### ISSUE THREAD

**Issue by menturion** [arvgta/ajaxify#199]
Hi,

many thanks for sharing this.

I am trying to get your Ajaxify script working. Main problem I am trying to solve is to prevent that all scripts (of the target page) are copied into the head instead of leaving them in the body. 

For instance, I have one container in which many scripts are referenced/loaded within the body:
```
<body>
// ...
<div id="loadscripts">
  // Scripts to be loaded
  <script async="" type="text/javascript" src="linktoscript-01"></script>
  <script async="" type="text/javascript" src="linktoscript-02"></script> 
  // ...
</div>
// ...
</body>
```

When loading the target page all references are copied to the head which causes many issues.
I have tried several different option settings with no success. 

Note: I also tried to declare the `"loadscripts"` container as content to be swapped.

Any suggestion is highly appreciated.

**Comment by arvgta** [comment]
Hi Menturion,

thanks for the interesting enquiry!

I believe the appending of these scripts is done in one central location only.  The good news is, that I suppose we could provide a switch for you to override the appending to the `head` and make Ajaxify append to the `body` instead...

The bad news is, that I won't be able to perform the change today - tomorrow maybe?

**Comment by menturion** [comment]
@arvgta 

Many thanks for your speedy reply and many many thanks for your offer! This would be great, but it isn't time critical.

**Comment by arvgta** [comment]
Sure, no problem!

Would that solve your use-case, if they are appended to the `body` instead?

I'll try tomorrow, or ask somebody to perform the change over night...

**Comment by menturion** [comment]
> Would that solve your use-case, if they are appended to the body instead?

I think that it will solve the current issues. Maybe, it would be helpful if there would be an option to specify a container (class or ID ) in which the target scripts would be appended.

**Comment by arvgta** [comment]
I am planning to change a simple:

    ...appendTo("head") //current code

to

    ...appendTo(appendsrc)

...with default:

    appendsrc: "head"

...which you can override with e.g.

    appendsrc: "body"

**Comment by menturion** [comment]
Sounds really good.

**Comment by arvgta** [comment]
Hi Menturion,

we had a look at the proposed change.   Unfortunately, we don't think that a universal change is possible.  The above fix was at the wrong place.  We have found the right place to tweak the code, but that would also affect external JS includes originating from the `head`, too, which is far from desirable.

Could you please describe, what issues you are experiencing, when the external JS's are appended to the head?

**Comment by menturion** [comment]
@arvgta 

Thanks for your reply.
I realized that (external) scripts are copied and not moved to the head so that each script is referenced twice in the DOM, i.e. within the head and at its original location (in my case within the `#loadscripts` container).

Tested without setting any options (i.e. w/ dafault settings only).

**Comment by menturion** [comment]
Changing the target location could be realized at line #442 (e.g. by `document.getElementById('loadscripts').appendChild(script)`). This would allow to define an option for specifying a selector of the target container. Or wouldn't this work?
https://github.com/arvgta/ajaxify/blob/cce2d99baf0cf0c7e776d2ec8461e9405b055df1/ajaxify.js#L442

**Comment by arvgta** [comment]
Thanks for both of your replies!

> I realized that (external) scripts are copied and not moved to the head so that each script is referenced twice in the DOM, i.e. within the head and at its original location (in my case within the #loadscripts container).

...we have to do something about that!  I suppose you have found this out by looking into the DOM?

Concerning the second issue, the line you highlighted is exactly the right one.  The catch is, that external script includes from the `head` are handled here as well, that's to say, they will be appended to the `body` or your container, too...

**Comment by arvgta** [comment]
Hi Menturion,

don't mean to be rude, but our first attempt to reproduce the bug about the double external `script` tags failed.

Furthermore, when checking the code, there seems to be a salient precaution on this line (currently 222):

```javascript
$c.find("script").remove(); //prevent double firing of scripts
```

**Comment by menturion** [comment]
@arvgta 

No problem. 
Nevertheless, the fact is that the scripts are referenced twice. Otherwise I wouldn't get issues of the kind ...

`Uncaught SyntaxError: Identifier 'DEBUG' has already been declared`

The `DEBUG` constant is defined only once in one file.

**Comment by arvgta** [comment]
Hi again,

my proposition:  Could you please send me a link to your (test) site to arvgta@gmail.com, if possible and I'll promise to treat it discretely...

**Comment by menturion** [comment]
@arvgta 

Many thanks for your offer! 
I'm testing your script in a local dev environment.

**Comment by arvgta** [comment]
Alright, I see, thanks...

It comes to my mind, that it has been reported quite often, that Ajaxify behaves different when set up locally.

How would you propose for us to go on?

**Comment by menturion** [comment]
Is there any reason why your script should behave differently, local vs live?

I will try to sort this out. Otherwise i would not be able to deploy your script.

**Comment by arvgta** [comment]
I don't know - it's just been reported quite often...

Best of luck with your own debugging efforts...

**Comment by menturion** [comment]
@arvgta 

The different behaviour occurs due to the expression `$((PK == "href" ? linkr : scrr).replace("!", $S))` does not match in line #445 https://github.com/arvgta/ajaxify/blob/3ecb0078308dd25497d6e3869a355d315a62849c/ajaxify.js#L445 in case of references that contain backslashes. This is quite often the case on windows systems when the links are generated dynamically w/o replacing the backslashes within the path strings.

Consequence is that the `_classAlways` assets won't be removed and thus referenced twice.

**Comment by arvgta** [comment]
Hi Menturion,

thanks for pinpointing the error!

Can you think of an appropriate fix, while we're at it?

**Comment by edito** [comment]
Hi @menturion 

you are saying that you use [delta-loading](https://4nf.org/delta-loading/) with `data-class="always"` with your dynamically added javascript files?

To explain a bit. CSS and JS behave different with that option.
CSS is simple, you add style to DOM it appears on website, you remove style from DOM it dissapears from website.

But with JS, once you add for example inline script with variable, it is permanently there until you refresh the page (even if you delete tag which contains link to the script from DOM). And that is causing you problems. So you have to be careful with handling javascript that way. 

You can check on [this link](https://javascript.info/garbage-collection) how javascript garbage collection works.

One solution for this problem is to use immediately-invoked function expressions inside script file that you are adding if you can, since they don't pollute global variable space. Here is the example of that:

```javascript
(function(){
    var foo = function() {
      // all your code here
    };
    foo();
    // ...
})();
// foo is unreachable here (it’s undefined)
```

Does this suits your needs?

**Comment by menturion** [comment]
@edito 

Many thanks for this reference but unfortunately this is no option in my case.

**Comment by menturion** [comment]
@arvgta 

> Can you think of an appropriate fix, while we're at it?

You need to escape the backslashes of all URLs that are part of a selector.
Just in case of the above mentioned line of code 

https://github.com/arvgta/ajaxify/blob/3ecb0078308dd25497d6e3869a355d315a62849c/ajaxify.js#L445

you would have to replace 

`$((PK == "href" ? linkr : scrr).replace("!", $S))`

by 

`$((PK == "href" ? linkr : scrr).replace("!", $.escapeSelector($S)))`

I just tested it and it seems to work. The scripts are referenced only once although the script hrefs contain backslashes.

**Comment by arvgta** [comment]
Hi Menturion,

that's great, thanks!  Interesting that there are no ready-made library functions for this purpose.

Could you please share the code of `jQuery.escapeSelector()`, as you've even tested it already...
([something with identical syntax](https://api.jquery.com/jQuery.escapeSelector/) actually already exists, but for a very different purpose)

**Comment by menturion** [comment]
@arvgta 

It's exactly the jQuery function `$.escapeSelector(...) `you referenced above (https://api.jquery.com/jQuery.escapeSelector/) and it's purpose is to escape special chars used within selectors. 

In the above line of code it is sufficient to just escape the URL string which is part of the selector. 

You need to do this everywhere in your script you are using hrefs/srcs as part of a selector in order to fix the potential different behaviour on windows OS (i.e. hrefs containg backslashes as described above).

**Comment by arvgta** [comment]
Thanks very much!  I guess that is a full specification of what has to be done - much appreciated!

I think there are not that many occurrences of hrefs/srcs in real jQuery selectors(at least I hope so)...

**Comment by arvgta** [comment]
Hi Menturion,

I just tried to employ `$.escapeSelector()` on two occasions (I think there are only two spots we have to tweak) and was surprised to get a hard error...

...because [$.escapeSelector()](https://api.jquery.com/jQuery.escapeSelector/) is only available from jQuery version 3.0.  However, I can't really demand that as a precondition.

The alternative might be to craft one's own little escaping function, that only replaces those Windows backslashes?
Might this be any good -> [Replace path with backslashes with forward slashes](https://stackoverflow.com/questions/13608564/path-with-backslashes-to-path-with-forward-slashes-javascript) ?

It's a pity, that `$.escapeSelector()` is not generally available, because that would have been a nice and lean fix...

**Comment by menturion** [comment]
@arvgta 

You can find the source here ...
https://github.com/jquery/jquery/blob/master/src/selector/escapeSelector.js

or e.g. here ...
https://gist.github.com/basselin/863c9f10e722b42feff6c382b15c6920

**Comment by arvgta** [comment]
Thanks Menturion,

would it be sufficient to replace the Windows backslashes with forward slashes - as described here:

- [Replace path with backslashes with forward slashes](https://stackoverflow.com/questions/13608564/path-with-backslashes-to-path-with-forward-slashes-javascript)

?

**Comment by menturion** [comment]
@arvgta 

I tried this and in my tests it didn't work. Then I used `$.escapeSelector()` instead and it worked instantly. I think I remember that the `$.escapeSelector()` escaped the forward slashes as well.

**Comment by arvgta** [comment]
Thanks, much appreciated.

All we need is the salient RegEx.

What would it look like to escape both?

(Sorry, I'm really bad at RegExes)

Once we have it, I'll employ it in those two occurrences...

**Comment by menturion** [comment]
See here https://gist.github.com/basselin/863c9f10e722b42feff6c382b15c6920#file-escapeselector-js

Have you already tried this as a function within your script? You can check if the function `$.escapeSelector()` exists in jQuery. If not use your replacement, i.e. the code above.

**Comment by arvgta** [comment]
I would prefer to employ a function, that is as lean as possible, in order not to bloat the code and also for run-time minimisation.

Yes, I had a look at:

- https://gist.github.com/basselin/863c9f10e722b42feff6c382b15c6920#file-escapeselector-js

...there is surely lots of logic in it, that we don't need...

An alternative to tweaking the code would be to simply educate the user in the documentation, that offline Windows systems behave differently.  If there is lots of code necessary, to support offline Windows, I would almost prefer to educate the user.

**Comment by edito** [comment]
```javascript
escape() // will not encode: @*/+

encodeURI() // will not encode: ~!@#$&*()=:/,;?+'

encodeURIComponent() // will not encode: ~!*()'
```

@menturion could you please test if one of these functions is good alternative for `$.escapeSelector()` in your case?

**Comment by menturion** [comment]
@edito 

These functions have different results and therefore are no replacements.

**Comment by menturion** [comment]
@arvgta 

The cost of parsing href strings is almost nothing, and I don't see much room for significant improvement of `$.escapeSelector()`. 

If you want to optimize your script with a minimum of overhead, you might consider recoding it in Vanilla JS.

**Comment by arvgta** [comment]
Hi Menturion,

did you know [these functions beforehand](https://github.com/arvgta/ajaxify/issues/199#issuecomment-617249779) and thus know that they are not appropriate?

With "Vanilla JS" I suppose you mean "normal JS" (there is a framework called "Vanilla JS" as well)
I would argue that without jQuery, the size of Ajaxify's code would be far larger?

**Comment by edito** [comment]
I've made RegEx variant:

```javascript
// escape path "/test\path/to\something/" to "\/test\\path\/to\\something\/"
path.replace(/(\/|\\)/g,"\\$1"); 
```
You can test matching on [this link](https://regex101.com/r/Kc9T2b/1).
RegEx should escape both, backward and forward slashes from variable. So, in our case, code should look like this:

```javascript
removeScript: $S => $((PK == "href" ? linkr : scrr).replace("!", $S.replace(/(\/|\\)/g,"\\$1"))).remove() //Remove script (stylesheet or external JS) from DOM
```

If we conclude that this doesn't work then there is something else in your URLs that should be escaped.

**Comment by arvgta** [comment]
Thanks for that Edin,

...exactly what I conceptually consider to be the best solution.

We agree that the Windows backslash is the first culprit - we only have to find the second one, too.

**Comment by menturion** [comment]
@edito 

I can confirm that it works for selectors based on links like "http://domain.com/static\script\test.js".

**Comment by edito** [comment]
> @edito
> 
> I can confirm that it works for selectors based on links like "[http://domain.com/static\script\test.js](http://domain.com/static%5Cscript%5Ctest.js)".

You are saying that it will solve your issue?

**Comment by menturion** [comment]
@edito 

It solves the general issue I reported which I fixed by using `$.escapeSelector()` (see above).

**Comment by arvgta** [comment]
Gentlemen, I've created [this pull-request](https://github.com/arvgta/ajaxify/pull/201/files), which I briefly tested against [this site](http://www.oeko-fakt.de/produkte/).
(It uses [alwayshints](http://www.oeko-fakt.de/wp-content/themes/oekoext/js/bob.js))

Edin, could you please test it in your regular fashion?

Do both of you agree, that we need it on this line additionally? ->

    if(PK == "href") return $(linki.replace("*", _esc(url))).appendTo("head"); //insert single stylesheet

**Comment by arvgta** [comment]
I tested a bit myself and there is a severe error - when traversing from:

- [This page](http://www.oeko-fakt.de/produkte/)

to 

- [this page](http://www.oeko-fakt.de/studien/)

You can see, that it is severely distorted...

not the case with the old file!

Edin:  I think the handling of the Windows backslash is alright, but I think the handling of the forward slash causes problems in normal environments?

I'm off for the evening...

**Comment by arvgta** [comment]
Performed a rollback and also [deleted the pull-request](https://github.com/arvgta/ajaxify/pull/201)...

I think we should take a step back to the specification stage:

As mentioned before, I feel positive about the escaping of Windows backslashes.
In my opinion, the errors we just saw when applying the fix were due to wrong escaping of "normal slashes" in our normal environment.

@menturion: Could you please state once more, what you want the "normal slashes" to be escaped to?  And what should happen in a normal environment?

Thanks once more to both of you!

**Comment by menturion** [comment]
@arvgta 

I just tested your commit you rolled back. It works fine w/o any problem.

**Comment by arvgta** [comment]
Hi Menturion,

I appreciate that - yes it seems to work in your Windows environment, but [not in a conventional one](https://github.com/arvgta/ajaxify/issues/199#issuecomment-617792897)?

(Please keep in mind, that I have already performed a rollback, so you won't be able to reproduce that error)

**Comment by menturion** [comment]
@arvgta 

> (Please keep in mind, that I have already performed a rollback, so you won't be able to reproduce that error)

:)  I changed the source manually according to your commit so I was able to test it ... against hrefs w/ and w/o backslashes. 

> I appreciate that - yes it seems to work in your Windows environment

JS depends on the JS engine of your browser, not the OS. So a different behaviour should not be expected. 

Maybe @edito can test your commit to be sure.

**Comment by arvgta** [comment]
Hi Menturion,

I really wouldn't mind if it worked ;)

However, [this bug I observed at the time](https://github.com/arvgta/ajaxify/issues/199#issuecomment-617792897) makes that commit not acceptable, from my point of view...

(I can imagine that the new logic caused a mismatch concerning conventional paths, making at least a stylesheet not load correctly)

May I ask once more - how would you like to have the normal slashes escaped?

**Comment by arvgta** [comment]
Just researched [differences in path handling on Windows systems](https://en.wikipedia.org/wiki/Path_(computing)#MS-DOS/Microsoft_Windows_style) a bit.

Everyone talks about escaping the backslash, but I can't find anything else?

Have we really tried replacing the Windows backslash only?  Something like ->

```javascript
path.replace(/\\/g,"/");
```

...a bit [like described here](https://stackoverflow.com/questions/42855606/replace-back-slash-with-forward-slash)....

**Comment by arvgta** [comment]
I have a suggestion:

If we should need a second replacement of a character, it might be simpler to chain a second `replace` after the first one, before we write a more complex and single RegEx...

**Comment by menturion** [comment]
@arvgta 

Maybe you should first test whether it is a regex issue at all by using the jQuery function `$.escapeSelector()` instead as suggested above. If this works it is a regex issue. If not there is another issue.

**Comment by arvgta** [comment]
Hi Menturion,

yes, we could even use [this function](https://gist.github.com/basselin/863c9f10e722b42feff6c382b15c6920#file-escapeselector-js) you suggested all the time, without checking for `$.escapeSelector()` at all, if the code is the same anyway.

What I don't like about this approach:

1. As mentioned before, the function is quite "heavy" in terms of code size -> 772 bytes
2. I don't understand the code within at all and don't like employing code, I don't understand

...and finally - [the description in the jQuery docs](https://api.jquery.com/jQuery.escapeSelector/) reads, as if it were geared at something completely different initially:

> This method is useful for situations where a class name or an ID contains characters that have a special meaning in CSS, such as the dot or the semicolon.
> 
> The method is essentially a shim for the CSS Working Group's CSS.escape() method. The main difference is that $.escapeSelector() can be reliably used in all of jQuery's supported browsers.

_Could we defer this whole issue to a later point in time?_

**Comment by menturion** [comment]
@arvgta 

Have you tested it using `$.escapeSelector()`?
Did it work in your environment?

**Comment by arvgta** [comment]
No, I haven't and it wouldn't matter that much, due to the reasons I just mentioned.

(I'm not convinced at a strategic level)

**Comment by arvgta** [comment]
Closing because the risks of changing this outweigh the benefits.

We will educate the user about these things in the docs for the time being...

**Comment by menturion** [comment]
@arvgta 

Your fix (https://github.com/arvgta/ajaxify/commit/4dacddb44aed609da9f9fb2165ec6ab5fd8295f9#diff-9395fad3c7372e47626ee7f892b752a2R436) does not work due to line #436:
`if(PK == "href") return $(linki.replace("*", _esc(url))).appendTo("head");`

**In line #436 (see your commit) you escaped the href of a CSS and added this escaped link to the DOM (to the head).** 

You only need to escape the URL in line #447: 
`removeScript: $S => $((PK == "href" ? linkr : scrr).replace("!", _esc($S))).remove()`

I guess you have assigned any always hint or `data-class="always"` to a CSS asset on your site(?).  In this case the selector in line #447 will not be able to match the previously changed/escaped href (see line #436) in the DOM.

So you just have to revert the change in line #436, i.e. remove the escape function `_esc()`, ... and keep the change in line #447.
